examples/gno.land/r/gnoland/ghverify/contract_test.gno

```diff
@@ -0,0 +1,99 @@
+package ghverify
+
+import (
+	"std"
+	"strings"
+	"testing"
+
+	"gno.land/p/demo/testutils"
+)
+
+func TestVerificationLifecycle(t *testing.T) {
+	defaultAddress := std.GetOrigCaller()
+	userAddress := std.Address(testutils.TestAddress("user"))
+
+	// Verify request returns no feeds.
+	result := GnorkleEntrypoint("request")
+	if result != "[]" {
+		t.Fatalf("expected empty request result, got %s", result)
+	}
+
+	// Make a verification request with the created user.
+	std.TestSetOrigCaller(userAddress)
+	RequestVerification("deelawn")
+
+	// A subsequent request from the same address should panic because there is
+	// already a feed with an ID of this user's address.
+	var errMsg string
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				errMsg = r.(string)
+			}
+		}()
+		RequestVerification("deelawn")
+	}()
+	if errMsg != "feed already exists" {
+		t.Fatalf("expected feed already exists, got %s", errMsg)
+	}
+
+	// Verify the request returns no feeds for this non-whitelisted user.
+	result = GnorkleEntrypoint("request")
+	if result != "[]" {
+		t.Fatalf("expected empty request result, got %s", result)
+	}
+
+	// Set the caller back to the whitelisted user and verify that the feed data
+	// returned matches what should have been created by the `RequestVerification`
+	// invocation.
+	std.TestSetOrigCaller(defaultAddress)
+	result = GnorkleEntrypoint("request")
+	expResult := `[{"id":"` + string(userAddress) + `","type":"0","value_type":"string","tasks":[{"gno_address":"` +
+		string(userAddress) + `","github_handle":"deelawn"}]}]`
+	if result != expResult {
+		t.Fatalf("expected request result %s, got %s", expResult, result)
+	}
+
+	// Try to trigger feed ingestion from the non-authorized user.
+	std.TestSetOrigCaller(userAddress)
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				errMsg = r.(string)
+			}
+		}()
+		GnorkleEntrypoint("ingest," + string(userAddress) + ",OK")
+	}()
```

Comment:
Shall we call directly to oracle.HandleMessage, and make that method to return an error to avoid these recoveries?

Reply:
I don't think so. The idea for these tests is to test the contract itself; no one can invoke a call to `oracle.HandleMessage` without going through a realm. But the good news is that I'll be able to remove these `recover`s since I am making changes to have the package return errors rather than panic, as per Morgan's comment above to be more aligned with what is published in the Effective Gno document.

Reply:
Actually what I said here is wrong. If we want to test the realm then we should expect a panic. Writing unit tests for the package itself won't have to deal with any kind of recovery because all functions and methods now return errors rather than panic.